carsus/io/cmfgen/base.py

```diff
@@ -0,0 +1,440 @@
+        for line in itertools.islice(f, start, stop):
+            for k in keys:
+                if k.lower() in line.lower():
+                    meta[k.strip('!')] = line.split()[0]
+
+    return meta
+
+
+def to_float(string):
+    """ String to float, taking care of Fortran 'D' values
+
+    Parameters
+    ----------
+    string : str
+
+    """
+    try:
+        value = float(string.replace('D', 'E'))
+
+    except ValueError:
+
+        if string == '1-.00':      # Bad value at MG/VIII/23oct02/phot_sm_3000 line 23340
+            value = 10.00
+
+        if string == '*********':  # Bad values at SUL/V/08jul99/phot_op.big lines 9255-9257
+            value = np.nan
+
+    return value
+
+
+class CMFGENEnergyLevelsParser(BaseParser):
+    """
+        Description
+        ----------
+        base : pandas.DataFrame
+        columns : list of str
+            (default value = ['Configuration', 'g', 'E(cm^-1)', 'eV', 'Hz 10^15', 'Lam(A)'])
+        meta : dict
+            Metadata parsed from file header.
+
+        Methods
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute.
+    """
+
+    keys = ['!Date',  # Metadata to parse from header. TODO: look for more keys
+            '!Format date',
+            '!Number of energy levels',
+            '!Ionization energy',
+            '!Screened nuclear charge',
+            '!Number of transitions',
+            ]
+
+    def load(self, fname):
+
+        meta = parse_header(fname, self.keys)
+        kwargs = {}
+        kwargs['header'] = None
+        kwargs['index_col'] = False
+        kwargs['sep'] = '\s+'
+        kwargs['skiprows'] = find_row(fname, "Number of transitions", num_row=True)
+
+        n = int(meta['Number of energy levels'])
+        kwargs['nrows'] = n
+
+        columns = ['Configuration', 'g', 'E(cm^-1)', 'eV', 'Hz 10^15', 'Lam(A)']
+
+        try:
+            df = pd.read_csv(fname, **kwargs, engine='python')
+
+        except pd.errors.EmptyDataError:
+            df = pd.DataFrame(columns=columns)
+            warnings.warn('Empty table')
+
+        # Assign column names by file content
+        if df.shape[1] == 10:
+            # Read column names and split them keeping one space (e.g. '10^15 Hz')
+            columns = find_row(fname, 'E(cm^-1)', "Lam").split('  ')
+            # Filter list elements containing empty strings
+            columns = [c for c in columns if c != '']
+            # Remove left spaces and newlines
+            columns = [c.rstrip().lstrip() for c in columns]
+            columns = ['Configuration'] + columns
+            df.columns = columns
+
+        elif df.shape[1] == 7:
+            df.columns = columns + ['#']
+            df = df.drop(columns=['#'])
+
+        elif df.shape[1] == 6:
+            df.columns = range(6)  # FIXME: These files don't have column names
```

Comment:
The columns should be ['Configuration', 'g', 'E(cm^-1)', 'Hz 10^15', 'Lam(A)', 'ID'].